prototype2/pubsub/src/lib.rs

```diff
@@ -0,0 +1,201 @@
+extern crate specs;
+extern crate itertools;
+
+use std::marker::PhantomData;
+use std::any::TypeId;
+use std::collections::HashMap;
+use std::sync::RwLockWriteGuard;
+use std::any::Any;
+use itertools::Itertools;
+
+
+/// ////////////////////////////////////////////////////////////////////////////////////////////////
+/// Publisher-Subscriber plugin for Specs ECS
+///
+/// The objective of this crate is to provide a safe API to emit and receive
+/// events via the Specs
+/// resource locking and unlocking mechanism.
+///
+/// TODO: Add more doc and overview
+/// ////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub trait PubSubStore {
+  fn read_publisher<'a, T: Clone + Any + Send + Sync>(&'a self,
+                                                      token: &PublisherToken<T>)
+                                                      -> Publisher<'a, T>;
+  fn read_subscriber<'a, T: Clone + Any + Send + Sync>(&'a self,
+                                                       token: &SubscriberToken<T>)
+                                                       -> Subscriber<'a, T>;
+
+  fn register_publisher<T: Clone + Any + Send + Sync>(&mut self) -> PublisherToken<T>;
+  fn register_subscriber<T: Clone + Any + Send + Sync>(&mut self) -> SubscriberToken<T>;
+}
+
+/// An indicator that the holder can access a pubsub resource and retrieve a
+/// provider corresponding
+/// to this type
+pub struct PublisherToken<T> {
+  phantom_data: PhantomData<T>,
+}
+
+/// An indicator pointing to a specific subscriber reader that the holder has
+/// access to.
+pub struct SubscriberToken<T> {
+  phantom_data: PhantomData<T>,
+  reader_idx: u32,
+}
+
+struct PubSubRegistry {
+  readers: HashMap<TypeId, u32>,
+}
+
+
+impl PubSubRegistry {
+  fn new() -> PubSubRegistry {
+    PubSubRegistry { readers: HashMap::new() }
+  }
+}
+
+impl PubSubStore for specs::World {
+  /// Attempt to retrieve the general publisher for this resource
+  fn read_publisher<'a, T: Clone + Any + Send + Sync>(&'a self,
+                                                      _: &PublisherToken<T>)
```

Comment:
Note, we literally don't need their token. The fact that they called this method with a token they can't instantiate themselves (easily) means they're ok.

This actually is useless though in the underlying model. Unguarded access to publishers is fine since it locks the readers. I wrote it this way with the intent for a future non-broadcast pubsub at some point.